### DISCUSSION TITLE
UserPrincipal doesn't have the key ID configured.

### DISCUSSION THREAD

**Discussion by vitmsadmin** [KelvinTegelaar/CIPP#1305] [General]
Have been using CIPP via Azure for a while and working well. 

However recently we came across an issue where it "Failed to retrieve tenants.  
We performed a permissions check and got this error.

**We could not connect to the API to retrieve the permissions. There might be a problem with the secure application model configuration. The returned error is: invalid_grant:AADSTS135010: UserPrincipal doesn't have the key ID configured. Trace ID: 38cbb353-dde0-4546-b9e5-32bc21079e00 Correlation ID: 025eae21-5b2c-4851-81ce-fa8c0506fcbe Timestamp: 2023-01-14 02:53:26Z**

We checked the updates which were a few versions out of sync so performed an update on GITHUB fork and on backend following standard process which has always worked.

However, we still cannot access tenants and get the same API permission error as shown above.

Following troubleshooting steps, we reset our ExchangeRefreshToken and RefreshToken which shows correctly as seen below
![image](https://user-images.githubusercontent.com/104887171/212447739-a4ceb5b6-2d6f-47b3-967c-182add71ca5d.png)

however, the issue remains.

How do we resolve this issue without having to rebuild the whole thing from scratch?

**Comment by vitmsadmin** [comment]
Results from troubleshooting using Token Testing Script

Standard Graph Refresh Token Test: 35 total tenants, with 0 failures
Exchange Refresh Token Test: 35 total tenants, with 0 failures
Write-Host 'All Tests Finished'

**Reply by stilles** [reply]
Also having the same issue crop up yesterday.

**Reply by stilles** [reply]
We ran through the "SAM Setup Wizard" and created a new application. Cleared the token cache after like it recommended and everything is back up and running today.
